locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/JcloudsLocation.java

```diff
@@ -2957,7 +2957,7 @@ String getHostnameAws(HostAndPort hostAndPort, LoginCredentials userCredentials,
                     "get public AWS hostname",
                     ImmutableList.of(
                             BashCommands.INSTALL_CURL,
-                            "echo `curl --silent --retry 20 http://169.254.169.254/latest/meta-data/public-hostname`; exit"));
+                            "echo `curl --silent --retry 20 http://169.254.169.254/latest/meta-data/"+(privateHost ? "local-hostname" : "public-hostname")+"`; exit"));
```

Comment:
thanks @tbouron, could we avoid the extra boolean parameter if we use something like
```
response=$(curl --write-out %{http_code} --silent --output /dev/null http://169.254.169.254/latest/meta-data/local-hostname)
if [ $(response) == "200" ] ; 
then
echo `curl --silent --retry 20 http://169.254.169.254/latest/meta-data/privateHost"`; exit"));
else
echo `curl --silent --retry 20 http://169.254.169.254/latest/meta-data/public-hostname"`; exit"));
```
or something similar?

Reply:
@andreaturli I don't think so. `http://169.254.169.254/latest/meta-data/local-hostname` 
 will always return some data and a 200 as AWS assign private IP/host for every single EC2 instance.

`http://169.254.169.254/latest/meta-data/public-hostname` could return a 404 (if the machine has been provisioned without an elastic IP, not entirely sure though) but in this situation, but that's not what we want

Reply:
not sure I follow, @tbouron -- I think it would be easier to read if we have 2 methods
`getPrivateAwsHostname` and `getPublicAwsHostname` what do you think?

Reply:
@andreaturli In you code example, `http://169.254.169.254/latest/meta-data/local-hostname` will always return a `200` (as an EC2 instance always as a private IP) so your if statement will always go one way which is not what we want here.

It would may be easier to read (although I'm not sure about that) but will also introduce some code duplication which I would be reluctant do to, especially in this huge file.

Reply:
sorry @tbouron, `String getHostnameAws(...)` is already very confusing, not mentioning the fact that is AWS specific which is another smell, IMHO.
Which hostname is going to return that method? 
I think we have the chance to improve this area of the code and I think another (boolean) parameter is not going to help in that respect. wdyt?

Reply:
If you are ok to add even more code + duplication then sure, no problem, I'll do that :D

Agree this is a code smell (which [has already been identified based on this comment](https://github.com/tbouron/brooklyn-server/blob/2455e054a62cbfff874c07c70f6cdf897f19769e/locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/JcloudsLocation.java#L2836-L2838)) but I don't have a magical solution to improve this particular area. If you have any idea, I'm opened to suggestions

Reply:
@tbouron if we can rely on jclouds for getting the private hostname `NodeMetadata.getHostname`, Brooklyn doesn’t need to calculate a new one especially for nodes created using JcloudsLocation. Notice `NodeMetadata.getHostname` is nullable so I think JcloudsLocation should manage that or assign the value to `host.addresses.private`

Reply:
@andreaturli I did what you suggested: using `NodeMetadata.getHostname` or reusing the `getPrivateHostnameGeneric()` method for private hostname. The public hostname is calculated as before.